src/axiom_use.rs

```diff
@@ -1,9 +1,138 @@
+            [Keyword(cmd), label, Keyword(avoids), axioms @ ..]
+                if cmd.as_ref(buf) == b"usage" && avoids.as_ref(buf) == b"avoids" =>
+            {
+                let stmt = label.value(buf);
+                let usage = self
+                    .axiom_use_map
+                    .get(stmt)
+                    .ok_or_else(|| vec![Diagnostic::UnknownLabel(label.full_span())])?;
+                let mut diags = vec![];
+                for cmd in axioms {
+                    let axiom = cmd.value(buf);
+                    if let Some(index) = self.axioms.iter().position(|&x| x == axiom) {
+                        if usage.has_bit(index) {
+                            // TODO possibly research the usage "path" leading to this error.
+                            diags.push(Diagnostic::UsageViolation(
+                                cmd.full_span(),
+                                stmt.into(),
+                                axiom.into(),
+                            ));
+                        }
+                    }
+                }
+                Err(diags)
+            }
+            _ => Ok(()),
+        }
+    }
+
+    /// Verify that all usage declarations are correct.
+    fn verify_usage(&'a mut self, sset: &'a SegmentSet) {
+        for sref in sset.segments(..) {
+            let mut j_commands: std::slice::Iter<'_, (i32, (crate::Span, Vec<CommandToken>))> =
+                sref.j_commands.iter();
+            for stmt in sref.range(..) {
+                match stmt.statement_type() {
+                    StatementType::AdditionalInfoComment => {
+                        while let Some(&(_, (_, ref args))) =
+                            j_commands.peeking_next(|i| i.0 == stmt.index)
+                        {
+                            if let Err(diags) = self.parse_command(sref, args) {
+                                for diag in diags {
+                                    self.result.0.push((stmt.address(), diag));
+                                }
+                            }
+                        }
+                    }
+                    StatementType::Provable => {
+                        let label = stmt.label();
+                        let mut usage = Bitset::new();
+                        let mut i = 1;
+                        loop {
+                            let tk = stmt.proof_slice_at(i);
+                            i += 1;
+                            if tk == b")" {
+                                break;
+                            }
+                            if let Some(usage2) = self.axiom_use_map.get(tk) {
+                                usage |= usage2
+                            }
+                            if i == stmt.proof_len() {
+                                break;
+                            }
+                        }
+                        self.axiom_use_map.insert(label, usage);
+                        eprintln!("Done");
```

Comment:
Was the `Done` output intended to be temporary? When run on set.mm this generates over 40,000 lines of output.

Reply:
Fixed, thanks!